### PR TITLE
feat(generate:typetests): Add per-package typetest entrypoint config

### PR DESCRIPTION
**Pull request by tylerbutler**
## Background

The type tests are primarily configured in the `typeValidation` field in package.json. However, the API level setting is only passed via a CLI flag, so when you make changes to the type tests for a particular project, you have to update the CLI command. It also means you can't use the package filtering flags to operate on a subset of packages with the CLI, which is confusing.

## Changes

The main change is a new typeValidation setting, `entrypoint`, that, when set, will be used as the "default" entrypoint for that package (default value is "legacy"). If the CLI `--entrypoint` flag is passed it is used instead of the default. (The `--level` flag is now a deprecated alias of the `--entrypoint` flag.)

I pre-applied the type test config changes needed in #22392 - they can be merged before the version of build-tools that uses them is released.

[AB#7875](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7875)

### CODE REVIEW

**Review by alexvy86 — COMMENTED** [review]
Without having looked at the whole PR, a couple of concerns:

- I thought "legacy" was orthogonal to API levels, i.e we (at some point) can have alpha, beta, public, and separately legacy, legacy alpha, legacy beta. So seeing legacy as the default `apiLevel` feels weird. I think I understand that in practice, today, it's kind of another api level, but if that's the case then I'd like to see a very deliberate acknowledgement of that and design around it.
- With `typetests:gen` now depending on `build:test`, did we check if we're affecting any packages that don't have a separate test build? Off the top of my head I don't recall which ones are like that but I know some exist. Maybe they don't require type tests and we'll be fine, but wanted to call it out.

**Comment by tylerbutler** [comment]
> I thought "legacy" was orthogonal to API levels, i.e we (at some point) can have alpha, beta, public, and separately legacy, legacy alpha, legacy beta. So seeing legacy as the default apiLevel feels weird. I think I understand that in practice, today, it's kind of another api level, but if that's the case then I'd like to see a very deliberate acknowledgement of that and design around it.

I don't think API Level is the best term anymore. For the purposes of type tests, these are "export paths" or "subsets of exported APIs with known names that correspond to a known export path." The string here tells us which types to load, and from an implementation perspective it should be able to handle arbitrary strings (e.g. if we had a package with /foo exports, then the tool, if provided the string "foo", would get the types from the /foo entrypoint -- if we allowed arbitrary string input, which we don't. But my point is that the strings are not special from a code perspective.

We also can generate multiple type test files if needed in the future by calling `generate typetests` multiple times. My original design was to have a type test per export path, but since there's duplication between alpha/beta/public that didn't make much sense in the end. But the infra is there for that to be done in the future if needed.

Maybe renaming the `level` flag to `entrypoint` would help clarify? That can be done without a breaking change with a flag alias.

> With typetests:gen now depending on build:test, did we check if we're affecting any packages that don't have a separate test build? Off the top of my head I don't recall which ones are like that but I know some exist. Maybe they don't require type tests and we'll be fine, but wanted to call it out.

Because of the way fluid-build works, if a package doesn't have a build:test script, then for that package, build:test is just ignored. In other words, for that package the typetests:gen task would run without waiting for build:test. The change _does_ impact the build though, which I call out in the "concerns" section of the PR description.

**Comment by tylerbutler** [comment]
> Maybe renaming the `level` flag to `entrypoint` would help clarify? That can be done without a breaking change with a flag alias.

I did this in the latest iteration.

**Comment by tylerbutler** [comment]
> > With typetests:gen now depending on build:test, did we check if we're affecting any packages that don't have a separate test build? Off the top of my head I don't recall which ones are like that but I know some exist. Maybe they don't require type tests and we'll be fine, but wanted to call it out.
> 
> Because of the way fluid-build works, if a package doesn't have a build:test script, then for that package, build:test is just ignored. In other words, for that package the typetests:gen task would run without waiting for build:test. The change _does_ impact the build though, which I call out in the "concerns" section of the PR description.

Correction here: typetests:gen does not depend on build:test - it's the other way around.

**Comment by tylerbutler** [comment]
Marking this as a draft again since #22312 needs to be merged first, after which this PR will change in size and scope.

File: build-tools/packages/build-cli/src/commands/generate/typetests.ts [hunk 1]
```diff
@@ -53,9 +54,12 @@ export default class GenerateTypetestsCommand extends PackageCommand<
...
 	static readonly flags = {
-		level: Flags.custom<ApiLevel>({
-			description: "What API level to generate tests for.",
-			default: ApiLevel.internal,
+		entrypoint: Flags.custom<ApiLevel>({
+			// Temporary alias for back-compat
+			aliases: ["level"],
+			deprecateAliases: true,
+			description:
+				'What entrypoint to generate tests for. Use "public" for the default entrypoint. If this flag is provided it will override the typeValidation.entrypoint setting in the package\'s package.json.',
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/commands/generate/typetests.ts:62]: Good job noting which takes priority when both are used. Also makes the other way to set it discoverable!

**Review by CraigMacomber — APPROVED** [review]

File: build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts [hunk 2]
```diff
@@ -18,18 +18,39 @@ export interface BrokenCompatSettings {
...
 	/**
 	 * If true, disables type test preparation and generation for the package.
+	 *
+	 * @defaultValue `false`
 	 */
 	disabled?: boolean;
 }
 
+export const defaultTypeValidationConfig: ITypeValidationConfig = {
+	entrypoint: "legacy",
```
Comment (alexvy86) [inline build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts:49]: Should "public" be the default? Not that avoiding breaking changes in legacy isn't important, but it seems like avoiding them in public should be _the_ top priority?

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts:49]: Legacy is a superset of public (I think - I need to double-check with Jason). Public covers the least amount of API surface.

Comment (jason-ha) [reply build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts:49]: legacy is indeed a superset of public (really for consumer convenience).
alpha is super set of beta and beta is superset of public as well.
alpha+beta+legacy only intersect with what is public. So, legacy does not include alpha or beta.

**Comment by tylerbutler** [comment]
Thanks for the review, @CraigMacomber!

File: build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts (hunk 2 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts:49]: Thanks for clarifying, Jason. Regardless of what we pick as the "default" here in code, ultimately each package explicitly states its own default, and I am planning to persist that in the package.json (explicit is better than implicit).

The current design proposal is that if we need to test legacy and alpha/beta, we generate multiple type test files by running the generator multiple times for each entrypoint we care about. There are alternatives we could consider if that proves too onerous.

Comment (jason-ha) [reply build-tools/packages/build-cli/src/typeValidator/typeValidatorConfig.ts:49]: If we want things to be explicit, then no default should be an option on the table.
I think oclif allows checks to say --entrypoint is required if --level isn't used.

**Pull request merged**